File: src/lib/ipc/bennu/format.ts
/**
 * Whole-file source transforms — the ones you ask for by pressing a key, not by putting the caret
 * on something.
 *
 * Both route on the file rather than on an engine: a language with a server is formatted by its
 * server, Java by Bennu's own formatter, and the editor calls one function either way. That is why
 * these do not live in `lsp.ts` — nothing here is a language-server concept.
 *
 * Both return **edits**, not replacement text: applying them through CodeMirror keeps the change in
 * the undo history as one step and the caret in place, and a transform that touched three lines does
 * not mark the whole file dirty.
 *
 * Same conventions as the rest of `ipc/bennu`: every call wraps its fields under `{ args: … }` and
 * every offset is a **UTF-8 byte offset**.
 */

import { bennu } from '../rpc';
import type { SourceEdit } from '$lib/types/bennu';

/**
 * Format the whole buffer, whichever engine knows the language: a server where there is one,
 * Bennu's own formatter for Java.
 *
 * An empty list means the file is already formatted — not that nothing could format it.
 *
 * Wire: `bennu_format` — `{ file, source, tab_size?, insert_spaces? }`.
 */
export function formatBuffer(
  file: string,
  source: string,
  tabSize?: number,
  insertSpaces?: boolean,
): Promise<SourceEdit[]> {
  return bennu('bennu_format', {
    args: { file, source, tab_size: tabSize, insert_spaces: insertSpaces },
  });
}

/**
 * Drop the imports the file does not use and put what is left in order — one edit over the whole
 * import block.
 *
 * Java only; anything else answers with nothing. Never collapses a package to a wildcard and never
 * adds an import — see the backend module for why. An empty list means there was nothing to change.
 *
 * Wire: `bennu_optimize_imports` — `{ file, source }`.
 */
export function optimizeImports(file: string, source: string): Promise<SourceEdit[]> {
  return bennu('bennu_optimize_imports', { args: { file, source } });
}

File: crates/products/bennu/runtime/be/src/format.rs
//! `format` domain — `bennu_format`: reformat a buffer, whoever knows how.
//!
//! One handler over both engines, because "format this file" is one question. A language with a
//! server is formatted by its server (`rustfmt` for Rust); Java is formatted by Bennu's own
//! formatter, since Bennu *is* the Java engine and there is no server to ask.
//!
//! The **style** splits the same way. A server reads the project's own file — `rustfmt.toml`,
//! `.clang-format`, `.prettierrc` — and all it is told from here is the indentation, because that
//! is what the editor is showing; duplicating the rest in Bennu's settings would create a second
//! truth that the CI does not read. Java has no such file, so its style *is* Bennu's settings.
//!
//! The editor calls this and never has to know which of the two answered — which is what stopped
//! Ctrl+Alt+L saying "no formatter for this file type" on the one language the product exists for.
//!
//! Edits rather than formatted text, in both directions: the editor applies them through CodeMirror
//! so the format lands in the undo history as one step and the caret keeps its place.

use bennu_core::prelude::BennuState;
use bennu_proto::prelude::SourceEdit;
use serde::Deserialize;

/// Args for [`bennu_format`].
#[derive(Deserialize)]
pub struct FormatArgs {
    pub file: String,
    pub source: String,
    /// Indent width. Absent → the editor's configured `indent_width`.
    #[serde(default)]
    pub tab_size: Option<u32>,
    /// Absent → spaces, matching Bennu's own editor default.
    #[serde(default)]
    pub insert_spaces: Option<bool>,
}

/// Reformat `source`. Empty when it is already formatted, or when nothing can format this file.
#[arbor_rpc::handler]
fn bennu_format(_ctx: &BennuState, args: FormatArgs) -> Result<Vec<SourceEdit>, String> {
    let cfg = bennu_core::config::load();
    let tab_size = args.tab_size.unwrap_or(cfg.indent_width).max(1) as usize;
    let spaces = args.insert_spaces.unwrap_or(!cfg.indent_with_tabs);

    if crate::intel::is_java_file(&args.file) {
        // The whole style comes from the profile. The two the caller may override are the two the
        // editor knows better than the config does — it is the surface the user is typing in, and
        // its indentation is what the formatter has to agree with.
        let style = bennu_intentions::prelude::FormatStyle {
            indent_width: tab_size,
            use_tabs: !spaces,
            max_blank_lines: cfg.java_max_blank_lines,
            indent_case_body: cfg.java_indent_case_body,
        };
        return Ok(bennu_intentions::prelude::format_edits(&args.source, style)
            .into_iter()
            .map(|e| SourceEdit {
                file: args.file.clone(),
                start: e.start,
                end: e.end,
                new_text: e.replacement,
            })
            .collect());
    }
    Ok(crate::lsp_route::format(&args.file, &args.source, tab_size as u32, spaces))
}
